### PR TITLE
OIDC: add friendly errors

### PR DESCRIPTION
**Pull request by kerryarchibald**
Fixes https://github.com/vector-im/element-web/issues/25665

Screenshot 2023-07-11 at 17 07 33

## Checklist

-   [ ] Tests written for new code (and old code if feasible)
-   [ ] Linter and other CI checks pass
-   [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * OIDC: add friendly errors ([\#11184](https://github.com/matrix-org/matrix-react-sdk/pull/11184)). Fixes vector-im/element-web#25665. Contributed by @kerryarchibald.

### CODE REVIEW

File: src/utils/oidc/error.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import { MatrixError } from "matrix-js-sdk/src/http-api";
+import { OidcError } from "matrix-js-sdk/src/oidc/error";
+
+import { _t } from "../../languageHandler";
+
+export enum OidcClientError {
+    InvalidQueryParameters = "Invalid query parameters for OIDC native login. `code` and `state` are required.",
+}
+
+export const getErrorMessage = (error: Error | MatrixError): string | ReactNode => {
```
Comment (richvdh) [inline src/utils/oidc/error.ts:27]: `MatrixError` is an `Error`, right? so this is equivalent to:

```suggestion
export const getErrorMessage = (error: Error): string | ReactNode => {
```

File: src/Lifecycle.ts [hunk 2]
```diff
@@ -66,6 +66,7 @@ import { OverwriteLoginPayload } from "./dispatcher/payloads/OverwriteLoginPaylo
 import { SdkContextClass } from "./contexts/SDKContext";
 import { messageForLoginError } from "./utils/ErrorUtils";
 import { completeOidcLogin } from "./utils/oidc/authorize";
+import { getErrorMessage } from "./utils/oidc/error";
```
Comment (richvdh) [inline src/Lifecycle.ts:69]: I feel like having a function called `getErrorMessage` imported into this file is pretty confusing, when that function only deals with OIDC errors.

Possibly the function should be called `getOidcErrorMessage`. Or we could do something like:

```suggestion
import * as OidcErrorUtils from "./utils/oidc/error";
```

... and then call `OidErrorUtils.getErrorMessage` below.

File: src/utils/oidc/error.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    switch (error.message) {
+        case OidcError.MissingOrInvalidStoredState:
+            return _t(
+                "We asked the browser to remember which homeserver you use to let you sign in, " +
+                    "but unfortunately your browser has forgotten it. Go to the sign in page and try again.",
+            );
+        case OidcClientError.InvalidQueryParameters:
+        case OidcError.CodeExchangeFailed:
+        case OidcError.InvalidBearerTokenResponse:
+        case OidcError.InvalidIdToken:
```
[the remark this answers is not in the document]
Comment (kerryarchibald) [reply src/utils/oidc/error.ts:37]: I'm not sure about 'relying' or 'expecting' - the purpose of the errors as far as js-sdk is concerned is to be informative in the logs. The friendly error messages are a nice-to-have with a default case. It's not expected that all errors passed to this function will have an `OidcError` or `OidcClientError` message.
I have added some documentation in js-sdk about `OidcError` use https://github.com/matrix-org/matrix-js-sdk/pull/3808

**Comment by kerryarchibald** [comment]
hey @richvdh, this PR is back from the dead 🧟‍♀️. Thanks for your review!

File: src/utils/oidc/error.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply src/utils/oidc/error.ts]: > I have added some documentation in js-sdk about OidcError use https://github.com/matrix-org/matrix-js-sdk/pull/3808

Yup that helps, thanks!

**Review by richvdh — APPROVED** [review]
lgtm!

**Pull request merged**